numba/core/typing/arraydecl.py

```diff
@@ -445,9 +445,12 @@ def resolve_view(self, ary, args, kws):
 
     @bound_function("array.astype")
     def resolve_astype(self, ary, args, kws):
+        from numba.core.errors import RequireLiteralValue
         from .npydecl import parse_dtype
         assert not kws
         dtype, = args
+        if dtype == types.unicode_type:
+            raise RequireLiteralValue("array.astype if dtype is a string it must be constant")
```

Comment:
This should probably be explicitly tested with a non-constant string and the error message checked cf. https://github.com/numba/numba/blob/a8e24a98d0cce7464043f5eca061ea1f45c3f537/numba/tests/test_array_methods.py#L491-L494

Reply:
I added an additional test to check for this. Since the previous infrastructure lowers dtype as a constant, I had to provide a new check and run function. Hopefully this infrastructure change is acceptable even if there is only 1 test that runs.

Reply:
Thanks, I've commented inline.